Button Kind testing

jcfilben opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
Adds testing for button kind. Test coverage for `StyledButtonKind.js` before:

**Statements:** 12% | 12/100 
**Branches:** 0% | 0/126
**Functions:** 0% | 0/19
**Lines:** 12.77% | 12/94

After:
**Statements:** 91% | 91/100
**Branches:** 76.98% | 97/126
**Functions:** 100% | 19/19
**Lines:** 90.43% | 85/94

Total testing coverage for the repository
Before:
Statements: 78.48% (4662/5940)
Branches: 69.23% (3707/5355)
Functions: 75.82% (1038/1369)
Lines: 79.72% (4494/5637)

After:
Statements: 80.28% (4743/5908)
Branches: 71.74% (3797/5293)
Functions: 77.34% (1058/1368)
Lines: 81.47% (4568/5607)

#### Where should the reviewer start?
Button-kind-test.js

#### What testing has been done on this PR?
I looked through the snapshots to make sure the tests were working and I looked at the coverage report to see line by line what was being covered for StyledButtonKind.js to maximize coverage without any unnecessary tests.

#### How should this be manually tested?
Run `yarn test --coverage` and look at the coverage report and snapshots.

#### Any background context you want to provide?

#### What are the relevant issues?
Closes #4169 

#### Screenshots (if appropriate)
Before:
<img width="1792" alt="Screen Shot 2020-06-12 at 12 02 43 PM" src="https://user-images.githubusercontent.com/54560994/84533132-c05a9f80-aca4-11ea-81ed-6b575371594b.png">

After:
<img width="1792" alt="Screen Shot 2020-06-12 at 12 00 19 PM" src="https://user-images.githubusercontent.com/54560994/84533152-c81a4400-aca4-11ea-9959-d5410cd28015.png">

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
Backward compatible

JoeButler7 commented on src/js/components/Button/__tests__/Button-test.js:
```diff
@@ -1,15 +1,31 @@
+  test('should have no accessibility violations', async () => {
+    const { container, getByText } = render(
+      <Grommet>
+        <Button label="Test" onClick={() => {}} />
```
would we be able to use `a11yTitle` here instead of label?

jcfilben replied:
I added the `a11yTitle` to the test but also left the `label="Test"` in there because they are doing two different things. The `a11yTitle` is for the aria-label and the `label="Test"` is for the label text that goes on the button

ShimiSun commented on src/js/components/Button/__tests__/Button-kind-test.js:
```diff
@@ -0,0 +1,240 @@
+
+  test('grommet theme', () => {
+    const { container } = render(
+      <Grommet theme={grommet}>
```
As discussed, let's use the button form hpe theme https://github.com/grommet/grommet-theme-hpe/blob/a2d7bbd8832f37c03e415a6cc6ec85178966a0bb/src/js/index.js#L243 instead of grommet theme.

ShimiSun commented on src/js/components/Button/__tests__/Button-kind-test.js:
```diff
@@ -0,0 +1,259 @@
+import { axe } from 'jest-axe';
+import { cleanup, fireEvent, render } from '@testing-library/react';
+import { Grommet, Button } from '../..';
+import { customTheme } from '../Button-kind-theme';
```
The filename "Button-kind-theme.js" should be renamed to lower case, we usually use title case for components. 
Also, I think we should change the name of the theme (customTheme) to match the name of the file, so maybe a name without hyphens.

ShimiSun commented on src/js/components/Button/__tests__/Button-kind-test.js:
```diff
@@ -0,0 +1,259 @@
+  test('custom theme', () => {
+    const { container } = render(
+      <Grommet theme={customTheme}>
+        <Button default />
```
We should also add tests for `<Button secondary /> ` and `<Button primary />` when the theme has the button.default and button.secondary defined.
It will be interesting to see if it will improve the coverage, it should.

jcfilben commented:
## Update:
- Changed the filename "Button-kind-theme.js" to "buttonKindTheme.js"
- Changed (customTheme) to (buttonKindTheme)
- Moved "buttonKindTheme.js" to `Button/__tests__/theme/buttonKindTheme.js` instead of `Button/buttonKindTheme.js`
- Added `"!src/js/components/**/__tests__/**/*.js"` to "collectCoverageFrom" to ignore files within folders inside of `components/**/__tests__` when collecting test coverage.
- Added `"components/(.*)/__tests__/(.*)/(.*).js"` to "testPathIgnorePatterns" to not count files within folders inside of `components/**/__tests__` as testing files.
- Added tests for the buttonKindTheme for `<Button secondary />` and `<Button primary />`.

JoeButler7 commented on src/js/components/Button/__tests__/Button-kind-test.js:
```diff
@@ -0,0 +1,277 @@
+          },
+        }}
+      >
+        <Button label="Test" plain hoverIndicator onClick={() => {}} />
```
Might it be better to test this with a `mouseOver` event instead of just setting `hoverIndicator` to true so that we can test the behavior as well?

jcfilben replied:
I don't think the mouseOver would change anything in the snapshot because the CSS for the hover in the snapshot is:
```
.c1:hover {
  background-color: rgba(221,221,221,0.4);
  color: #444444;
}
```
The CSS and the shapshot wouldn't change if there was a mouseOver or not. Although the user would see a change, the code in the snapshot wouldn't be any different.

jcfilben commented:
Moved to https://github.com/grommet/grommet/pull/4209 because of issues merging with master
